### PR TITLE
[MLIR] add inferReturnTypeComponents for mhlo reduce op

### PR DESCRIPTION
**Pull request by CaramelFc**
Hi, This commit add the `inferReturnTypeComponets` for `mhlo.reduce`

### CODE REVIEW

**Comment by cheshire** [comment]
@sherhut

File: tensorflow/compiler/xla/mlir_hlo/lib/Dialect/mhlo/IR/hlo_ops.cc [hunk 1]
```diff
@@ -5785,6 +5784,30 @@ LogicalResult ReduceOp::reifyReturnTypeShapes(
...
 
+LogicalResult ReduceOp::inferReturnTypeComponents(
+    MLIRContext* context, Optional<Location> location, ValueShapeRange operands,
+    DictionaryAttr attributes, RegionRange regions,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
+
+  ReduceOp::Adaptor adaptor(operands, attributes, regions);
+  auto inputs = adaptor.inputs();
+  for (Value input : inputs) {
+    if(input == nullptr){
```
Comment (sherhut) [inline tensorflow/compiler/xla/mlir_hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:5795]: Drop this conditional. input cannot be `nulllptr` here. All `inputs` of a reduce operation need to be defined ssa values.

File: tensorflow/compiler/xla/mlir_hlo/lib/Dialect/mhlo/IR/hlo_ops.cc [hunk 2]
```diff
@@ -5785,6 +5784,30 @@ LogicalResult ReduceOp::reifyReturnTypeShapes(
...
+      return failure();
+    }
+    auto tensorType = getReduceResultType(input.getType(), adaptor.dimensions());
```
Comment (sherhut) [inline tensorflow/compiler/xla/mlir_hlo/lib/Dialect/mhlo/IR/hlo_ops.cc:5798]: Return early. For example

```
auto tensorType = getReduceResultType(input.getType(), adaptor.dimensions());
if (!tensorType.hasRank()) return failure();
```

File: tensorflow/compiler/xla/mlir_hlo/tests/Dialect/mhlo/mhlo_infer_shape_type_methods.mlir [hunk 3]
```diff
@@ -415,4 +415,31 @@ func.func @add_bounds_unranked(
...
-}
\ No newline at end of file
+}
+
+// -----
+
+// CHECK-LABEL: @reduce
+func.func @reduce(%arg0: tensor<?x4xf32>) -> tensor<?xindex> {
+  %0 = mhlo.constant dense<-0.000000e+00> : tensor<f32>
+  %1 = mhlo.reduce(%arg0 init: %0) applies mhlo.add across dimensions = [1] : (tensor<?x4xf32>, tensor<f32>) -> tensor<?xf32>
```
Comment (sherhut) [inline tensorflow/compiler/xla/mlir_hlo/tests/Dialect/mhlo/mhlo_infer_shape_type_methods.mlir:425]: Could you make this a variadic reduce to increase coverage? Ideally with a 2-d result that is only partially dynamic.

**Review by sherhut — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/mlir_hlo/lib/Dialect/mhlo/IR/hlo_ops.cc (hunk 2 above)
Comment (CaramelFc) [reply tensorflow/compiler/xla/mlir_hlo/lib/Dialect/mhlo/IR/hlo_ops.cc]: done

File: tensorflow/compiler/xla/mlir_hlo/tests/Dialect/mhlo/mhlo_infer_shape_type_methods.mlir [hunk 4]
```diff
@@ -415,4 +415,62 @@ func.func @add_bounds_unranked(
...
+// CHECK-LABEL: @reduce
+func.func @reduce(%arg0: tensor<?x4x12xf32>, %arg1: tensor<?x4x12xi32>) -> tensor<?xindex>{
+  %0 = mhlo.constant dense<-0.000000e+00> : tensor<f32>
+  %1 = mhlo.constant dense<1> : tensor<i32>
+  %2:2 = "mhlo.reduce"(%arg0, %arg1, %0, %1) ({
+  ^bb0(%arg4: tensor<f32>, %arg5: tensor<i32>, %arg6: tensor<f32>, %arg7: tensor<i32>):
+    %4 = mhlo.add %arg4, %arg6 : tensor<f32>
+    %5 = mhlo.add %arg5, %arg7 : tensor<i32>
+    "mhlo.return"(%4, %5) : (tensor<f32>, tensor<i32>) -> ()
+  }) {dimensions = dense<[1, 2]> : tensor<2xi64>}
```
Comment (sherhut) [inline tensorflow/compiler/xla/mlir_hlo/tests/Dialect/mhlo/mhlo_infer_shape_type_methods.mlir:431]: If you make the reduction dimensions here `[2]`, then the result of the reduction will have a partially dynamic shape and as a result, we should see this turn up in the return_type_components.

**Review by sherhut — CHANGES_REQUESTED** [review]
Thanks for the changes. A small change to the test and this is good to go.

File: tensorflow/compiler/xla/mlir_hlo/tests/Dialect/mhlo/mhlo_infer_shape_type_methods.mlir (hunk 4 above)
Comment (CaramelFc) [reply tensorflow/compiler/xla/mlir_hlo/tests/Dialect/mhlo/mhlo_infer_shape_type_methods.mlir:431]: add a test that results the  partially dynamic shape

**Review by sherhut — APPROVED** [review]

**Comment by gbaned** [comment]
@CaramelFc Can you please resolve conflicts? Thank you!

**Comment by CaramelFc** [comment]
> @CaramelFc Can you please resolve conflicts? Thank you!

hi, conflicts were resolved

**Review by sherhut — APPROVED** [review]

File: tensorflow/compiler/xla/mlir_hlo/tests/Dialect/mhlo/mhlo_infer_shape_type_methods.mlir [hunk 5]
```diff
@@ -415,4 +415,81 @@ func.func @add_bounds_unranked(
...
+  // CHECK: return %[[SHAPE1]], %[[SHAPE2]] : tensor<1xindex>, tensor<1xindex>
+  %0 = mhlo.constant dense<-0.000000e+00> : tensor<f32>
+  %1 = mhlo.constant dense<1> : tensor<i32>
+  %2:2 = mhlo.reduce(%arg0 init: %0), (%arg1 init: %1) across dimensions = [1, 2] : (tensor<?x4x12xf32>, tensor<?x4x12xi32>, tensor<f32>, tensor<i32>) -> (tensor<?xf32>, tensor<?xi32>)
+   reducer(%arg2: tensor<f32>, %arg4: tensor<f32>) (%arg3: tensor<i32>, %arg5: tensor<i32>)  {
+    %4 = mhlo.add %arg2, %arg4 : tensor<f32>
+    %5 = mhlo.add %arg3, %arg5 : tensor<i32>
+    mhlo.return %4, %5 : tensor<f32>, tensor<i32>
+  }
+  %3 = "mhlo_test.reify_return_type_shapes"(%2#0) : (tensor<?xf32>) -> tensor<1xindex>
```
Comment (sherhut) [inline tensorflow/compiler/xla/mlir_hlo/tests/Dialect/mhlo/mhlo_infer_shape_type_methods.mlir:477]: I realized this does not work. `"mhlo_test.reify_return_type_shapes"` returns the reified shape components for all results of the producer of the passed value. So for the example here, you should only have something like

```
%3:2 = "mhlo_test.reify_return_type_shapes"(%2#0) : (tensor<?xf32>) -> (tensor<1xindex>, tensor<1xindex>)
return %3:0, %3:1 : tensor<1xindex>, tensor<1xindex>
```

See also the [definition of the test pass](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/xla/mlir_hlo/lib/Dialect/mhlo/transforms/test_infer_shaped_type_pass.cc;l=106) for details.

**Review by sherhut — CHANGES_REQUESTED** [review]

**Review by sherhut — APPROVED** [review]
Thanks for the fixes.

**Comment by gbaned** [comment]
@CaramelFc Can you please resolve conflicts? Thank you!

**Comment by CaramelFc** [comment]
> Tried to land this again but the change is still failing tests. The failures are in
> 
> ```
> tensorflow/compiler/xla/mlir_hlo/tests:Dialect/mhlo/hlo-legalize-to-linalg.mlir.test	FAILED
> tensorflow/compiler/xla/mlir_hlo/tests:Dialect/mhlo/verifier_reduce_op.mlir.test
> ```
> 
> Have you run these tests?

hi, this failed test was fixed, and the commit rebased the master

**Review by sherhut — APPROVED** [review]

**Comment by sherhut** [comment]
There is a merge conflict blocking this change. Please rebase again.

**Comment by sherhut** [comment]
Different merge conflict now. There seems to be quite some contention on mhlo :(

**Comment by gbaned** [comment]
@CaramelFc Can you please resolve conflicts? Thank you!

**Comment by gbaned** [comment]
Hi @CaramelFc Any update on this PR? Please. Thank you!

**Comment by gbaned** [comment]
@CaramelFc  I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen! Thank you!

**Pull request closed without merging**